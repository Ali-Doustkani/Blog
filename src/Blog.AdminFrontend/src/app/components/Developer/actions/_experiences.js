import uuid from 'uuid/v1'
import { update } from '../../../utils/fn'

const createExperience = () => ({
   id: uuid(),
   company: '',
   position: '',
   startDate: today(),
   endDate: '',
   content: ''
})

const addExperience = state => ({
   ...state,
   experiences: [...state.experiences, createExperience()]
})

const deleteExperience = (state, action) => ({
   ...state,
   experiences: state.experiences.filter(x => x.id !== action.id)
})

const updateExperience = (state, action) => ({
   ...state,
   experiences: state.experiences.map(update(action.change))
})

function today() {
   const now = new Date()
   return `${now.getFullYear()}-${String(now.getMonth() + 1).padStart(2, 0)}-${String(
      now.getDate()
   ).padStart(2, 0)}`
}

export { addExperience, updateExperience, deleteExperience }
