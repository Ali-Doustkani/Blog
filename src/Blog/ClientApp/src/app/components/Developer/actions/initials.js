const STATUS = {
   LOADING: 1,
   SAVING: 2,
   IDLE: 3,
   STOPPED: 4
}

const initialState = {
   status: STATUS.LOADING,
   errorMessage: '',
   summary: '',
   summaryErrors: [],
   skills: '',
   skillsErrors: [],
   experiences: [],
   experiencesErrors: [],
   sideProjects: [],
   sideProjectErrors: []
}

export { initialState, STATUS }
