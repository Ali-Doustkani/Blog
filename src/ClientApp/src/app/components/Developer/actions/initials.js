const STATUS = {
   LOADING: 1,
   SAVING: 2,
   IDLE: 3,
   STOPPED: 4
}

const initialState = {
   status: STATUS.LOADING,
   errorMessage: '',
   summary: '',
   skills: '',
   experiences: [],
   sideProjects: [],
   educations: []
}

export { initialState, STATUS }
