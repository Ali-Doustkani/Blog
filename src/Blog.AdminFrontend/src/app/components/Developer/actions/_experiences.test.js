import { addExperience, deleteExperience, updateExperience } from './_experiences'
import { initialState } from './initials'
import uuid from 'uuid/v1'

jest.mock('uuid/v1')

afterEach(jest.restoreAllMocks)

it('adds a new valid experience', () => {
   uuid.mockReturnValue(123)
   jest.spyOn(global, 'Date').mockImplementation(() => ({
      getFullYear: () => 2012,
      getMonth: () => 4,
      getDate: () => 20
   }))
   const init = {
      experiences: []
   }
   const newState = addExperience(init)

   expect(newState.experiences).toEqual([
      {
         id: 123,
         company: '',
         position: '',
         startDate: '2012-05-20',
         endDate: '',
         content: ''
      }
   ])
})

it('deletes an existing experience', () => {
   const initial = {
      experiences: [{ id: 1 }, { id: 2 }]
   }

   const newState = deleteExperience(initial, { id: 1 })

   expect(newState.experiences).toEqual([{ id: 2 }])
})

it('updates an experience', () => {
   uuid.mockReturnValue(1)
   const init = {
      experiences: [
         {
            id: 1,
            company: 'Lodgify',
            position: 'C# Developer',
            startDate: '2010-01-01',
            endDate: '2011-01-01',
            content: 'DESC'
         }
      ]
   }

   let state = updateExperience(init, { change: { id: 1, company: 'Parmis' } })
   state = updateExperience(state, { change: { id: 1, position: 'JS' } })
   state = updateExperience(state, { change: { id: 1, startDate: '2013-01-02' } })
   state = updateExperience(state, { change: { id: 1, endDate: '2015-01-01' } })
   state = updateExperience(state, { change: { id: 1, content: 'Description' } })

   expect(state.experiences).toEqual([
      {
         id: 1,
         company: 'Parmis',
         position: 'JS',
         startDate: '2013-01-02',
         endDate: '2015-01-01',
         content: 'Description'
      }
   ])
})

it('keeps the state intact', () => {
   uuid.mockReturnValueOnce(1)
   uuid.mockReturnValueOnce(2)
   jest.spyOn(global, 'Date').mockImplementation(() => ({
      getFullYear: () => 2019,
      getMonth: () => 6,
      getDate: () => 2
   }))

   let state = addExperience(initialState)
   state = addExperience(state)
   state = updateExperience(state, { change: { id: 2, company: 'Lodgify' } })
   state = updateExperience(state, { change: { id: 2, position: 'C# Developer' } })
   state = deleteExperience(state, { id: 2 })

   expect(state).toMatchSnapshot()
})
