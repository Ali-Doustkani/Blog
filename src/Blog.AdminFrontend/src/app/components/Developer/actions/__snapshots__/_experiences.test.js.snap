// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`keeps the state intact 1`] = `
Object {
  "errorMessage": "",
  "experiences": Array [
    Object {
      "company": "",
      "content": "",
      "endDate": "",
      "id": 1,
      "position": "",
      "startDate": "2019-07-02",
    },
  ],
  "experiencesErrors": Array [],
  "sideProjectErrors": Array [],
  "sideProjects": Array [],
  "skills": "",
  "skillsErrors": Array [],
  "status": 1,
  "summary": "",
  "summaryErrors": Array [],
}
`;
